### PR TITLE
Upload logs and artifacts to GCS

### PR DESCRIPTION
**Pull request by briandealwis**
For #501: as is, this PR uploads the build artifacts into GCS.  It currently uploads to a bucket in a project I created named `gcloud-eclipse-testing`.  The PR archives the following artifacts:

  - contents of any `surefire-reports` directories (for surefire logs)
  - contents of any `screenshots` directories: these any snapshots taken by SWTBot's `SWTUtils.captureScreenshot()` or `SWTBotJUnit4ClassRunner` (on test failure)
  - on success, uploads `gcp-repo/target/repository` from the Oxygen build

The bucket layout looks like:
  - each build is placed in a directory _travis/{build-number}_ (e.g., `travis/11027`)
  - the p2 repository is at `repository` (e.g., `travis/11027/repository`)
  - the test logs and screenshots are then under _{job-number}-{os}_, and preserves the layout of the git repository (e.g., `travis/11027/11027.3-linux/plugins/com.google.cloud.tools.eclipse.welcome.test/target/surefire-reports/com.google.cloud.tools.eclipse.welcome.PluginXmlTest.txt`)

It works out to about 18MB per build at the moment.

You can see it in action at:

    https://console.cloud.google.com/storage/browser/gcloud-eclipse-testing/

Everything is currently public, and we upload results on pull requests and normal builds.  I have a lifecycle for objects of 7 days.  Note that I've tweaked the layout since the first two builds; `11027` and later are good exemplars.

I've put instructions for configuring the GCS bucket and configuring the service keys in an accompanying README.

### CODE REVIEW

**Comment by chanseokoh** [comment]
```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.0.0:compile (default-compile) on project com.google.cloud.tools.eclipse.integration.appengine: Compilation failure: Compilation failure: 
[ERROR] /home/travis/build/GoogleCloudPlatform/google-cloud-eclipse/plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/BaseProjectTest.java:[72] 
[ERROR] 	String fileName = SWTBotPreferences.SCREENSHOTS_DIR + "/" + "timeout-" + project.getName()
[ERROR] 	       ^^^^^^^^
[ERROR] Duplicate local variable fileName
[ERROR] 1 problem (1 error)
```

**Comment by briandealwis** [comment]
Sorry, I realized I hadn't actually verified that a screenshot was uploaded too.

File: .travis.yml [hunk 1]
```diff
@@ -13,34 +17,51 @@ jdk: oraclejdk8
...
     # ECLIPSE_TARGET is the targeted platform: mars is the default target
-    # JAVA_RUNTIME specifies the execution runtime for tests; must match an <id> in .travis-toolchains.xml
+    # JAVA_RUNTIME specifies the execution runtime for tests; must match an <id> in .travisci/toolchains.xml
     # SWT_GTK3=0 to force GTK2 on Mars as its GTK3 support is buggy
     - JAVA_RUNTIME=JavaSE-1.7 SWT_GTK3=0
     - JAVA_RUNTIME=JavaSE-1.8 ECLIPSE_TARGET=neon
     - JAVA_RUNTIME=JavaSE-1.8 ECLIPSE_TARGET=oxygen MAVEN_FLAGS=-Pjacoco
   global:
-    - PATH=$PWD/google-cloud-sdk/bin:$PATH
+    - GCLOUDSDK=$HOME/google-cloud-sdk
```
Comment (chanseokoh) [inline .travis.yml]: I prefer having underscores: `GCLOUD_SDK` or `CLOUD_SDK_HOME`.

File: .travis.yml [hunk 2]
```diff
@@ -13,34 +17,51 @@ jdk: oraclejdk8
...
+    - PATH=$GCLOUDSDK/bin:$PATH
     # Our maven build typically reports about 512m
     - MAVEN_OPTS=-Xmx700m
     - CLOUDSDK_CORE_DISABLE_USAGE_REPORTING=true
     - DISPLAY=:99.0
+    - GCSBUILDBUCKET=gs://gcloud-eclipse-testing
```
Comment (chanseokoh) [inline .travis.yml]: ditto

File: .travis.yml [hunk 3]
```diff
@@ -13,34 +17,51 @@ jdk: oraclejdk8
...
+  # Copy surefire-reports and screenshots
+  - mkdir build-results && .travisci/copy-test-results.sh plugins build-results
+  - $GCLOUDSDK/bin/gsutil -m -q
+      cp -r -z 'txt,xml,html' -a public-read
+        build-results/*
+        $GCSBUILDBUCKET/travis/${TRAVIS_BUILD_NUMBER}/${TRAVIS_JOB_NUMBER}-${TRAVIS_OS_NAME}/
+  # Copy build results (if successul; for only one element of build matrix)
+  - if [ -d gcp-repo/target/repository -a -d build/jacoco/target ]; then
+      $GCLOUDSDK/bin/gsutil -m -q
+        cp -r -z 'txt,xml,html' -a public-read
```
Comment (chanseokoh) [inline .travis.yml]: File extensions don't seem right.

File: .travis.yml [hunk 4]
```diff
@@ -13,34 +17,51 @@ jdk: oraclejdk8
...
+      -in .travisci/travis-service-account.json.enc -out /tmp/travis-service-account.json -d
+  - $GCLOUDSDK/bin/gcloud auth activate-service-account --key-file=/tmp/travis-service-account.json
+  # Copy surefire-reports and screenshots
+  - mkdir build-results && .travisci/copy-test-results.sh plugins build-results
+  - $GCLOUDSDK/bin/gsutil -m -q
+      cp -r -z 'txt,xml,html' -a public-read
+        build-results/*
+        $GCSBUILDBUCKET/travis/${TRAVIS_BUILD_NUMBER}/${TRAVIS_JOB_NUMBER}-${TRAVIS_OS_NAME}/
+  # Copy build results (if successul; for only one element of build matrix)
+  - if [ -d gcp-repo/target/repository -a -d build/jacoco/target ]; then
```
Comment (chanseokoh) [inline .travis.yml]: Q: jacoco is to verify that the Maven run was OK?

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/BaseProjectTest.java [hunk 5]
```diff
@@ -61,6 +61,11 @@ public void tearDown() {
 
       // ensure there are no jobs
       SwtBotWorkbenchActions.waitForProjects(bot, project);
+      {
+        String fileName = SWTBotPreferences.SCREENSHOTS_DIR + "/" + "post-test-" + project.getName()
+            + "." + SWTBotPreferences.SCREENSHOT_FORMAT.toLowerCase();
+        SWTUtils.captureScreenshot(fileName);
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/BaseProjectTest.java]: Do you intend to always capture in a normal situation, or is this for a temporary testing?

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/FlexMavenPackagedProjectStagingDelegateTest.java [hunk 6]
```diff
@@ -31,9 +31,12 @@
 import org.eclipse.core.runtime.IPath;
 import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.swtbot.swt.finder.junit.SWTBotJunit4ClassRunner;
 import org.junit.Assume;
 import org.junit.Test;
+import org.junit.runner.RunWith;
 
+@RunWith(SWTBotJunit4ClassRunner.class)
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/FlexMavenPackagedProjectStagingDelegateTest.java]: Is this necessary? If so, I wonder why?

File: .travisci/README.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+```
+$ gsutil lifecycle set .travisci/gcs.lifecycle.json gs://BUCKET
+```
+
+### Configuring the Service Account Key
+
+  1. Create a key for the service account from the [Cloud
+     Console](https://console.cloud.google.com/iam-admin/serviceaccounts)
+     with role _Storage > Storage Object Creator_.
```
Comment (elharo) [inline .travisci/README.md]: I think you also need to verify that "Enable G Suite Domain-wide Delegation" is checked, though it should be by default.

File: .travisci/README.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+     Check the _Furnish a new private key_.  Click _Create_.
+     This will produce a JSON file.
+
+  2. Follow Travis-CI's [instructions for automatically encrypting
+     files](https://docs.travis-ci.com/user/encrypting-files/).
+     Replace the `$encrypted_XXXX_*` values in the `.travis.yml`.
+
```
Comment (elharo) [inline .travisci/README.md]: should there be a step 3 to commit this to our repo?

**Comment by briandealwis** [comment]
Verified images are uploaded; I've now backed out that change:

https://console.cloud.google.com/storage/browser/gcloud-eclipse-testing/travis/11033/11033.3-linux/plugins/com.google.cloud.tools.eclipse.integration.appengine/screenshots/?project=gcloud-eclipse-testing

They're JPEGs, which seem ok, but we can tweak them with `-Dorg.eclipse.swtbot.screenshots.format=png`.

File: .travis.yml (hunk 4 above)
Comment (briandealwis) [reply .travis.yml]: We only perform the jacoco on a single element of our build matrix.  So this test prevents multiple uploads of what should be the same thing.

File: .travis.yml (hunk 3 above)
Comment (briandealwis) [reply .travis.yml]: How so?  These are the extensions for `gsutil` to compress.  We have the following extensions in what we upload:

  - html: a stub in our `gcp-repo` that redirects to http://cloud.google.com/eclipse
  - index: the `p2.index` file, which is tiny
  - jar
  - jpeg
  - txt: the surefire test reports
  - xml: the surefire test properties
  - xml.xz: the `gcp-repo` p2 metadata files

It's futile to compress `jar`, `xml.xz`, `zip`, `jpeg`, `png`.  We could add `index`.

File: .travisci/README.md (hunk 7 above)
Comment (briandealwis) [reply .travisci/README.md]: Don't we want it *unchecked*?  Does this service need to access user's data?

> In enterprise applications you may want to programmatically access a user's data without any manual authorization on their part. In G Suite domains, the domain administrator can grant third-party applications with domain-wide access to its users' data — this is referred as domain-wide delegation of authority. To delegate authority this way, domain administrators can use service accounts with OAuth 2.0.

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/FlexMavenPackagedProjectStagingDelegateTest.java (hunk 6 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/FlexMavenPackagedProjectStagingDelegateTest.java]: It takes a screenshot on failed tests.  We already use this annotation on most of our integration tests.

File: .travis.yml (hunk 3 above)
Comment (chanseokoh) [reply .travis.yml]: Ah, I didn't know `-z` is for compression.

File: .travisci/README.md (hunk 7 above)
Comment (elharo) [reply .travisci/README.md]: I read that as, if unchecked then a user will have to go through an oAuth screen before access is granted. I.e. a pure service account is not enough. User action is required.

I could be misreading it.

Comment (briandealwis) [reply .travisci/README.md]: I verified that my service account has this option unchecked.  Pantheon shows the following text:

screen shot 2017-12-15 at 3 07 38 pm

**Comment by briandealwis** [comment]
PTAL @elharo: I updated #501 with [remaining work items](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/501#issuecomment-356154304).

File: .travis.yml [hunk 9]
```diff
@@ -13,34 +17,52 @@ jdk: oraclejdk8
...
   # test req'd as we don't run coverage on all elements of the matrix
   - if [ -d build/jacoco/target ]; then bash <(curl -s https://codecov.io/bash); fi
+after_script:
+  # Upload build reports to our Google Cloud Storage bucket
+  - openssl aes-256-cbc -K $encrypted_42b011b17512_key -iv $encrypted_42b011b17512_iv
+      -in .travisci/travis-service-account.json.enc -out /tmp/travis-service-account.json -d
+  - $CLOUD_SDK_HOME/bin/gcloud auth activate-service-account --key-file=/tmp/travis-service-account.json
+  # Copy surefire-reports and screenshots
+  - mkdir build-results && .travisci/copy-test-results.sh plugins build-results
+  ###- $CLOUD_SDK_HOME/bin/gsutil -m -q
```
Comment (elharo) [inline .travis.yml]: Are these comments still relevant? Can we delete these lines?

File: .travis.yml [hunk 10]
```diff
@@ -13,34 +17,52 @@ jdk: oraclejdk8
...
     - JAVA_RUNTIME=JavaSE-1.8 ECLIPSE_TARGET=oxygen MAVEN_FLAGS=-Pjacoco
   global:
-    - PATH=$PWD/google-cloud-sdk/bin:$PATH
+    - CLOUD_SDK_HOME=$HOME/google-cloud-sdk
+    - PATH=$CLOUD_SDK_HOME/bin:$PATH
     # Our maven build typically reports about 512m
     - MAVEN_OPTS=-Xmx700m
     - CLOUDSDK_CORE_DISABLE_USAGE_REPORTING=true
     - DISPLAY=:99.0
+    - GCS_BUILD_BUCKET=gs://gcloud-eclipse-testing
```
Comment (elharo) [inline .travis.yml]: gs://travis_artifacts

File: .travisci/README.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+```
+$ gsutil lifecycle set .travisci/gcs.lifecycle.json gs://BUCKET
+```
+
+### Configuring the Service Account Key
+
+  1. Create a key for the service account from the [Cloud
+     Console](https://console.cloud.google.com/iam-admin/serviceaccounts)
+     with role _Storage > Storage Object Creator_.
+     Check the _Furnish a new private key_.  Click _Create_.
```
Comment (elharo) [inline .travisci/README.md]: I'm going to send you a new unencrypted key for the travis_artifacts bucket over email.

File: .travis.yml (hunk 9 above)
Comment (briandealwis) [reply .travis.yml]: They're relevant: they're the final bits to upload the results.

File: .travisci/README.md (hunk 11 above)
Comment (briandealwis) [reply .travisci/README.md]: thx

**Comment by briandealwis** [comment]
`DeployPropertyPageTest` failed with an SWTBot `TimeoutException`, but no screenshots were taken as the test class wasn't annotated with `@RunWith(SWTBotJunit4ClassRunner.class)`.
```
testPropertyPageTitle_standardProject(com.google.cloud.tools.eclipse.integration.appengine.DeployPropertyPageTest)  Time elapsed: 49.572 sec  <<< ERROR!
org.eclipse.swtbot.swt.finder.widgets.TimeoutException: Timeout after: 30000 ms.: Shell Resource - foo/src/main/java/HelloAppEngine.java - Eclipse Platform did not close
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.waitUntil(SWTBotFactory.java:437)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.waitUntil(SWTBotFactory.java:411)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.waitUntil(SWTBotFactory.java:399)
	at com.google.cloud.tools.eclipse.swtbot.SwtBotTestingUtilities.waitUntilShellIsNotActive(SwtBotTestingUtilities.java:121)
	at com.google.cloud.tools.eclipse.swtbot.SwtBotTestingUtilities.performAndWaitForWindowChange(SwtBotTestingUtilities.java:86)
	at com.google.cloud.tools.eclipse.swtbot.SwtBotProjectActions.openProjectProperties(SwtBotProjectActions.java:195)
	at com.google.cloud.tools.eclipse.integration.appengine.DeployPropertyPageTest.testPropertyPageTitle_standardProject(DeployPropertyPageTest.java:45)
```

**Comment by briandealwis** [comment]
Hmm, the built artifacts aren't being uploaded. Uploading the build results succeeds, but the repo fails.  I wonder if I need to add a trailing '/'.
```
$CLOUD_SDK_HOME/bin/gsutil -m -q cp -r -z 'txt,xml,html' -a public-read build-results/* $GCS_BUILD_BUCKET/travis/${TRAVIS_BUILD_NUMBER}/${TRAVIS_JOB_NUMBER}-${TRAVIS_OS_NAME}/

travis_time:end:033ab5ee:start=1515555641124738842,finish=1515555648067215339,duration=6942476497
[0Ktravis_fold:end:after_script.4
[0Ktravis_fold:start:after_script.5
[0Ktravis_time:start:012601a4
[0K$ if [ -d gcp-repo/target/repository -a -d build/jacoco/target ]; then $CLOUD_SDK_HOME/bin/gsutil -m -q cp -r -z 'txt,xml,html' -a public-read gcp-repo/target/repository $GCS_BUILD_BUCKET/travis/${TRAVIS_BUILD_NUMBER}/repository; fi
AccessDeniedException: 403 travis-artifacts@gcloud-for-eclipse-testing.google.com.iam.gserviceaccount.com does not have storage.objects.list access to travis_artifacts.
```

**Comment by chanseokoh** [comment]
Doesn't it saying 403 AccessDeniedException?

**Comment by briandealwis** [comment]
It is, but look at the remainder:
```
travis-artifacts@gcloud-for-eclipse-testing.google.com.iam.gserviceaccount.com does not have storage.objects.list access to travis_artifacts.
```
But the previous call to `gsutil` succeeded.

**Comment by chanseokoh** [comment]
`travis-artifacts@gcloud-for-eclipse-testing.google.com.iam.gserviceaccount.com` is allowed to create new GCS objects but not to read. If the trailing '/' doesn't fix it, we may need to give the read permission.

**Comment by briandealwis** [comment]
Rewriting the repository-upload command (https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2673/commits/770560009239e54921da5f473b7723bdc4f126fb) seems to have done the trick: the `gcp-repo` was uploaded and available.

   https://storage.googleapis.com/travis_artifacts/travis/11290/repository/

(Do you think we can grant anonymous users `storage.objects.get` access to `/travis` and below?)

One thing to note: I had to restart the Oxygen build job due to a spurious test failure (#2714).  Since restarting keeps the same build job number (in this case, `11290.3`), the build logs are re-uploaded and which fails:
```
travis_time:end:21360a46:start=1515610702531151263,finish=1515610703081504325,duration=550353062
[0Ktravis_fold:end:after_script.3
[0Ktravis_fold:start:after_script.4
[0Ktravis_time:start:3c3ef970
[0K$ $CLOUD_SDK_HOME/bin/gsutil -m -q cp -r -z 'txt,xml,html' -a public-read build-results/* $GCS_BUILD_BUCKET/travis/${TRAVIS_BUILD_NUMBER}/${TRAVIS_JOB_NUMBER}-${TRAVIS_OS_NAME}/
AccessDeniedException: 403 travis-artifacts@gcloud-for-eclipse-testing.google.com.iam.gserviceaccount.com does not have storage.objects.delete access to travis_artifacts/travis/11290/11290.3-linux/plugins/com.google.cloud.tools.eclipse.appengine.deploy.test/target/surefire-reports/com.google.cloud.tools.eclipse.appengine.deploy.flex.FlexWarStagingDelegateTest.txt.
AccessDeniedException: 403 travis-artifacts@gcloud-for-eclipse-testing.google.com.iam.gserviceaccount.com does not have storage.objects.delete access to travis_artifacts/travis/11290/11290.3-linux/plugins/com.google.cloud.tools.eclipse.appengine.compat.test/target/surefire-reports/TEST-com.google.cloud.tools.eclipse.appengine.compat.gpe.GpeClassicMigratorTest.xml.
[…many similar lines deleted…]
AccessDeniedException: 403 travis-artifacts@gcloud-for-eclipse-testing.google.com.iam.gserviceaccount.com does not have storage.objects.delete access to travis_artifacts/travis/11290/11290.3-linux/plugins/com.google.cloud.tools.eclipse.util.test/target/surefire-reports/com.google.cloud.tools.eclipse.util.ArtifactRetrieverTest.txt.
CommandException: 562 files/objects could not be transferred.
```
Fortunately it doesn't fail the build.  I don't know if we want to grant `storage.objects.delete` to the travis service account?

**Comment by briandealwis** [comment]
Oh, to see the surefire logs, etc, in the absence of anonymous `storage.objects.get` permission, use:

   https://console.cloud.google.com/storage/browser/travis_artifacts/travis/

**Comment by briandealwis** [comment]
PTAL: this is ready for commit.  Any other tweaks are permission issues.

**Comment by chanseokoh** [comment]
> Oh, to see the surefire logs, etc, in the absence of anonymous storage.objects.get permission, use:

So I guess we don't need to grant `storage.objects.get` to anonymous users?

> Fortunately it doesn't fail the build. I don't know if we want to grant storage.objects.delete to the travis service account?

I will give it full permission. I don't see a problem.

**Comment by briandealwis** [comment]
Anonymous users will be able to fetch files if they know the path; they won't be able to get directory listings.  So they can install a build from a …/repository/ URL since the file names are derived or encoded in the p2 artifacts and metadata files.  But they cannot navigate through the build logs and screenshots.

> I will give it full permission. I don't see a problem.

I think the only danger is a malicious PR could `gsutil rm -r gs://travis_artifacts`.  But this bucket is ephemeral.  Anyways, the permission did allow overwriting the build results.

**Comment by briandealwis** [comment]
> Just want to confirm, no one except travis can decrypt the key and use it, right?

Although that is correct, the decrypted service-account file is written out and an attacker could modify the .travis.yml to execute an arbitrary command using the decrypted service-account file.

**Comment by briandealwis** [comment]
And I suppose they could upload the decrypted service-account file somewhere else.

**Comment by briandealwis** [comment]
Verified that `TRAVIS_SECURE_ENV_VARS=true` on PRs from team members and 
 `TRAVIS_SECURE_ENV_VARS=false` on PRs from non-team members.

**Comment by briandealwis** [comment]
I changed the repository uploading to use `gsutil rsync`, but the logs show:

> WARNING: gsutil rsync uses hashes when modification time is not available at
> both the source and destination. Your crcmod installation isn't using the
> module's C extension, so checksumming will run very slowly. If this is your
> first rsync since updating gsutil, this rsync can take significantly longer than
> usual. For help installing the extension, please see "gsutil help crcmod".

I'll revert to using `gsutil cp`.

**Comment by briandealwis** [comment]
Failures are due to #2711.

**Pull request merged**